layer4/routes.go

```diff
@@ -184,5 +183,8 @@ func (routes RouteList) Compile(logger *zap.Logger, matchingTimeout time.Duratio
 				}
 			}
 		}
+
+		logger.Error("matching connection", zap.String("remote", cx.RemoteAddr().String()), zap.Error(errors.New("number of prefetch calls exhausted")))
+		return nil
```

Comment:
I wonder if we should return the error here. But this is also probably fine. Any thoughts?

Reply:
It's that way because I wanted to differentiate errors during matching vs errors from the handlers. "matching connection" vs. "handling connection". But I agree it looks bad/wrong.  
We could swap the positions. Log "matching connection" in [server.go#L168](https://github.com/mholt/caddy-l4/blob/master/layer4/server.go#L168) and "handling connection" in [routes.go#L173](https://github.com/mholt/caddy-l4/blob/master/layer4/routes.go#L173). Then the `HandlerFunc` in `Compile` could return errors directly everywhere.  
But the logic for only logging with warning level for `ErrDeadlineExceeded` should probably be kept in the compiled function. Otherwise it must be duplicated to all other places that call `Handle` on the compiled route.  Like the listener wrapper or the subroute handler.

Reply:
Yeah, so it's in a function that acts as a handler representing the entire route, which is both matching and handling. That's a good question, I guess, whether we should treat their errors separately or differently. A matcher should only return an error if a decision (true/false) couldn't be reached; that should be rare, and notable, since it indicates a problem with the connection. Maybe we should treat their errors equally?

Reply:
Sorry for the delayed response. For now, I would leave it as it is. I think the different messages are useful to differentiate errors during the routing/matching phase and caddy actually knowing what to do with the connection. Since errors during matching are likely caused by clients sending garbage, disconnecting early or configuration problems. But errors during handing likely indicate real problems for real/valid users.